### ISSUE TITLE
false positive `reportGeneralTypeIssues` for list of typing.Protocol

### ISSUE THREAD

**Issue by ovsds** [microsoft/pyright#3892]
Hello, having next code sample i recieve `reportGeneralTypeIssues` on `example2` function.

```
import dataclasses
import typing


class TestProtocol(typing.Protocol):
    value: str


@dataclasses.dataclass
class Test:
    value: str


def test_func(test: list[TestProtocol]) -> None:
    print(test)


def example1() -> None:
    test_func(test=[Test(value="test")])


def example2() -> None:
    test = [Test(value="test")]
    test_func(test=test)
```
Pyright error message:
```
Argument of type "list[Test]" cannot be assigned to parameter "test" of type "list[TestProtocol]" in function "test_func"
    TypeVar "_T@list" is invariant
      "Test" is incompatible with protocol "TestProtocol" (reportGeneralTypeIssues)
```

Notably if i change prototype of test_func for `TestProtocol` instead of `list[TestProtocol]`, everything works just fine:
```
def test_func2(test: TestProtocol) -> None:
    print(test)


def example3() -> None:
    test_func2(test=Test(value="test"))


def example4() -> None:
    test = Test(value="test")
    test_func2(test=test)
```
Thanks for your work! :)

**Comment by erictraut** [comment]
This isn't a false positive. Pyright's behavior is correct here. Mypy emits the same error in this case.

The problem is that the type parameter for `list` is invariant, so the type must match exactly, but you are passing a subtype (because `Test` is a subtype of `TestProtocol`).

You have a couple of options:
1. Use `Sequence` rather than `list`. This works because the type parameter for Sequence is covariant.
2. In `example2`, use an explicit type declaration for the variable `test`. This will override the normal type inference for the list expression. (For details about type inference refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md).)
```python
    test: list[TestProtocol] = [Test(value="test")]
```

**Comment by ovsds** [comment]
Thanks for your reply, now i got it. Have a good day.
